File: payment-core/go-services/internal/rbac/permify_adapter.go
package rbac

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// PermifyConfig holds Permify connection configuration
type PermifyConfig struct {
	URL       string `json:"url" yaml:"url"`
	TenantID  string `json:"tenant_id" yaml:"tenantId"`
	APIKey    string `json:"api_key" yaml:"apiKey"`
}

// Entity represents a Permify entity
type Entity struct {
	Type string `json:"type"`
	ID   string `json:"id"`
}

// Subject represents a Permify subject
type Subject struct {
	Type     string `json:"type"`
	ID       string `json:"id"`
	Relation string `json:"relation,omitempty"`
}

// Tuple represents a relationship tuple
type Tuple struct {
	Entity   Entity  `json:"entity"`
	Relation string  `json:"relation"`
	Subject  Subject `json:"subject"`
}

// CheckRequest represents a permission check request
type CheckRequest struct {
	TenantID   string            `json:"tenant_id"`
	Entity     Entity            `json:"entity"`
	Permission string            `json:"permission"`
	Subject    Subject           `json:"subject"`
	Metadata   map[string]string `json:"metadata,omitempty"`
}

// CheckResponse represents a permission check response
type CheckResponse struct {
	Can            bool   `json:"can"`
	RemainingDepth int    `json:"remaining_depth,omitempty"`
}

// WriteRequest represents a relationship write request
type WriteRequest struct {
	TenantID string  `json:"tenant_id"`
	Tuples   []Tuple `json:"tuples"`
}

// WriteResponse represents a relationship write response
type WriteResponse struct {
	SnapToken string `json:"snap_token"`
}

// DeleteRequest represents a relationship delete request
type DeleteRequest struct {
	TenantID string  `json:"tenant_id"`
	Tuples   []Tuple `json:"tuples"`
}

// LookupEntityRequest represents an entity lookup request
type LookupEntityRequest struct {
	TenantID   string            `json:"tenant_id"`
	EntityType string            `json:"entity_type"`
	Permission string            `json:"permission"`
	Subject    Subject           `json:"subject"`
	Metadata   map[string]string `json:"metadata,omitempty"`
}

// LookupEntityResponse represents an entity lookup response
type LookupEntityResponse struct {
	EntityIDs []string `json:"entity_ids"`
}

// PermifyAdapter provides Permify authorization operations
type PermifyAdapter struct {
	config     PermifyConfig
	httpClient *http.Client
}

// NewPermifyAdapter creates a new Permify adapter
func NewPermifyAdapter(config PermifyConfig) *PermifyAdapter {
	return &PermifyAdapter{
		config: config,
		httpClient: &http.Client{
			Timeout: 10 * time.Second,
		},
	}
}

// Check performs a permission check
func (p *PermifyAdapter) Check(ctx context.Context, req CheckRequest) (*CheckResponse, error) {
	if req.TenantID == "" {
		req.TenantID = p.config.TenantID
	}

	url := fmt.Sprintf("%s/v1/tenants/%s/permissions/check", p.config.URL, req.TenantID)

	body, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal check request: %w", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create check request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	if p.config.APIKey != "" {
		httpReq.Header.Set("Authorization", "Bearer "+p.config.APIKey)
	}

	resp, err := p.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to perform check: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("check failed: %s - %s", resp.Status, string(respBody))
	}

	var checkResp CheckResponse
	if err := json.NewDecoder(resp.Body).Decode(&checkResp); err != nil {
		return nil, fmt.Errorf("failed to decode check response: %w", err)
	}

	return &checkResp, nil
}

// WriteRelationships writes relationship tuples
func (p *PermifyAdapter) WriteRelationships(ctx context.Context, tuples []Tuple) (*WriteResponse, error) {
	url := fmt.Sprintf("%s/v1/tenants/%s/relationships/write", p.config.URL, p.config.TenantID)

	req := WriteRequest{
		TenantID: p.config.TenantID,
		Tuples:   tuples,
	}

	body, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal write request: %w", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create write request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	if p.config.APIKey != "" {
		httpReq.Header.Set("Authorization", "Bearer "+p.config.APIKey)
	}

	resp, err := p.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to write relationships: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("write failed: %s - %s", resp.Status, string(respBody))
	}

	var writeResp WriteResponse
	if err := json.NewDecoder(resp.Body).Decode(&writeResp); err != nil {
		return nil, fmt.Errorf("failed to decode write response: %w", err)
	}

	return &writeResp, nil
}

// DeleteRelationships deletes relationship tuples
func (p *PermifyAdapter) DeleteRelationships(ctx context.Context, tuples []Tuple) error {
	url := fmt.Sprintf("%s/v1/tenants/%s/relationships/delete", p.config.URL, p.config.TenantID)

	req := DeleteRequest{
		TenantID: p.config.TenantID,
		Tuples:   tuples,
	}

	body, err := json.Marshal(req)
	if err != nil {
		return fmt.Errorf("failed to marshal delete request: %w", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("failed to create delete request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	if p.config.APIKey != "" {
		httpReq.Header.Set("Authorization", "Bearer "+p.config.APIKey)
	}

	resp, err := p.httpClient.Do(httpReq)
	if err != nil {
		return fmt.Errorf("failed to delete relationships: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("delete failed: %s - %s", resp.Status, string(respBody))
	}

	return nil
}

// LookupEntities looks up entities a subject has permission on
func (p *PermifyAdapter) LookupEntities(ctx context.Context, req LookupEntityRequest) (*LookupEntityResponse, error) {
	if req.TenantID == "" {
		req.TenantID = p.config.TenantID
	}

	url := fmt.Sprintf("%s/v1/tenants/%s/permissions/lookup-entity", p.config.URL, req.TenantID)

	body, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal lookup request: %w", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("failed to create lookup request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")
	if p.config.APIKey != "" {
		httpReq.Header.Set("Authorization", "Bearer "+p.config.APIKey)
	}

	resp, err := p.httpClient.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to lookup entities: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		respBody, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("lookup failed: %s - %s", resp.Status, string(respBody))
	}

	var lookupResp LookupEntityResponse
	if err := json.NewDecoder(resp.Body).Decode(&lookupResp); err != nil {
		return nil, fmt.Errorf("failed to decode lookup response: %w", err)
	}

	return &lookupResp, nil
}

// Payment Switch specific helper methods

// CanAccessMerchant checks if a user can access a merchant
func (p *PermifyAdapter) CanAccessMerchant(ctx context.Context, userID, merchantID, permission string) (bool, error) {
	resp, err := p.Check(ctx, CheckRequest{
		Entity:     Entity{Type: "merchant", ID: merchantID},
		Permission: permission,
		Subject:    Subject{Type: "user", ID: userID},
	})
	if err != nil {
		return false, err
	}
	return resp.Can, nil
}

// CanAccessTransaction checks if a user can access a transaction
func (p *PermifyAdapter) CanAccessTransaction(ctx context.Context, userID, transactionID, permission string) (bool, error) {
	resp, err := p.Check(ctx, CheckRequest{
		Entity:     Entity{Type: "transaction", ID: transactionID},
		Permission: permission,
		Subject:    Subject{Type: "user", ID: userID},
	})
	if err != nil {
		return false, err
	}
	return resp.Can, nil
}

// AssignUserToMerchant assigns a user to a merchant with a role
func (p *PermifyAdapter) AssignUserToMerchant(ctx context.Context, userID, merchantID, role string) error {
	_, err := p.WriteRelationships(ctx, []Tuple{
		{
			Entity:   Entity{Type: "merchant", ID: merchantID},
			Relation: role,
			Subject:  Subject{Type: "user", ID: userID},
		},
	})
	return err
}

// RemoveUserFromMerchant removes a user from a merchant
func (p *PermifyAdapter) RemoveUserFromMerchant(ctx context.Context, userID, merchantID, role string) error {
	return p.DeleteRelationships(ctx, []Tuple{
		{
			Entity:   Entity{Type: "merchant", ID: merchantID},
			Relation: role,
			Subject:  Subject{Type: "user", ID: userID},
		},
	})
}

// GetUserMerchants gets all merchants a user has access to
func (p *PermifyAdapter) GetUserMerchants(ctx context.Context, userID, permission string) ([]string, error) {
	resp, err := p.LookupEntities(ctx, LookupEntityRequest{
		EntityType: "merchant",
		Permission: permission,
		Subject:    Subject{Type: "user", ID: userID},
	})
	if err != nil {
		return nil, err
	}
	return resp.EntityIDs, nil
}

File: payment-core/security/auth/permify_adapter.py
"""
Permify Authorization Adapter for Payment Switch Platform

Provides fine-grained RBAC/ABAC authorization using Permify.
"""

import httpx
from dataclasses import dataclass, field
from typing import Optional, List, Dict, Any
import logging

logger = logging.getLogger(__name__)


@dataclass
class PermifyConfig:
    """Permify connection configuration"""
    url: str
    tenant_id: str = "payment-switch"
    api_key: Optional[str] = None
    verify_ssl: bool = True


@dataclass
class Entity:
    """Permify entity"""
    type: str
    id: str


@dataclass
class Subject:
    """Permify subject"""
    type: str
    id: str
    relation: Optional[str] = None


@dataclass
class Tuple:
    """Relationship tuple"""
    entity: Entity
    relation: str
    subject: Subject

    def to_dict(self) -> Dict[str, Any]:
        result = {
            "entity": {"type": self.entity.type, "id": self.entity.id},
            "relation": self.relation,
            "subject": {"type": self.subject.type, "id": self.subject.id},
        }
        if self.subject.relation:
            result["subject"]["relation"] = self.subject.relation
        return result


@dataclass
class CheckRequest:
    """Permission check request"""
    entity: Entity
    permission: str
    subject: Subject
    tenant_id: Optional[str] = None
    metadata: Optional[Dict[str, str]] = None


@dataclass
class CheckResponse:
    """Permission check response"""
    can: bool
    remaining_depth: Optional[int] = None


@dataclass
class WriteResponse:
    """Relationship write response"""
    snap_token: str


@dataclass
class LookupEntityRequest:
    """Entity lookup request"""
    entity_type: str
    permission: str
    subject: Subject
    tenant_id: Optional[str] = None
    metadata: Optional[Dict[str, str]] = None


@dataclass
class LookupEntityResponse:
    """Entity lookup response"""
    entity_ids: List[str]


class PermifyAdapter:
    """Permify authorization adapter"""

    def __init__(self, config: PermifyConfig):
        self.config = config
        self._client: Optional[httpx.AsyncClient] = None

    @property
    def client(self) -> httpx.AsyncClient:
        if self._client is None:
            headers = {"Content-Type": "application/json"}
            if self.config.api_key:
                headers["Authorization"] = f"Bearer {self.config.api_key}"
            self._client = httpx.AsyncClient(
                timeout=10.0,
                verify=self.config.verify_ssl,
                headers=headers
            )
        return self._client

    async def close(self):
        """Close the HTTP client"""
        if self._client:
            await self._client.aclose()
            self._client = None

    async def check(self, req: CheckRequest) -> CheckResponse:
        """Perform a permission check"""
        tenant_id = req.tenant_id or self.config.tenant_id
        url = f"{self.config.url}/v1/tenants/{tenant_id}/permissions/check"

        payload = {
            "tenant_id": tenant_id,
            "entity": {"type": req.entity.type, "id": req.entity.id},
            "permission": req.permission,
            "subject": {"type": req.subject.type, "id": req.subject.id},
        }
        if req.subject.relation:
            payload["subject"]["relation"] = req.subject.relation
        if req.metadata:
            payload["metadata"] = req.metadata

        response = await self.client.post(url, json=payload)

        if response.status_code != 200:
            logger.error(f"Permission check failed: {response.text}")
            raise Exception(f"Permission check failed: {response.status_code}")

        result = response.json()
        return CheckResponse(
            can=result.get("can", False),
            remaining_depth=result.get("remaining_depth"),
        )

    async def write_relationships(self, tuples: List[Tuple]) -> WriteResponse:
        """Write relationship tuples"""
        url = f"{self.config.url}/v1/tenants/{self.config.tenant_id}/relationships/write"

        payload = {
            "tenant_id": self.config.tenant_id,
            "tuples": [t.to_dict() for t in tuples],
        }

        response = await self.client.post(url, json=payload)

        if response.status_code != 200:
            logger.error(f"Write relationships failed: {response.text}")
            raise Exception(f"Write relationships failed: {response.status_code}")

        result = response.json()
        return WriteResponse(snap_token=result.get("snap_token", ""))

    async def delete_relationships(self, tuples: List[Tuple]) -> None:
        """Delete relationship tuples"""
        url = f"{self.config.url}/v1/tenants/{self.config.tenant_id}/relationships/delete"

        payload = {
            "tenant_id": self.config.tenant_id,
            "tuples": [t.to_dict() for t in tuples],
        }

        response = await self.client.post(url, json=payload)

        if response.status_code != 200:
            logger.error(f"Delete relationships failed: {response.text}")
            raise Exception(f"Delete relationships failed: {response.status_code}")

    async def lookup_entities(self, req: LookupEntityRequest) -> LookupEntityResponse:
        """Look up entities a subject has permission on"""
        tenant_id = req.tenant_id or self.config.tenant_id
        url = f"{self.config.url}/v1/tenants/{tenant_id}/permissions/lookup-entity"

        payload = {
            "tenant_id": tenant_id,
            "entity_type": req.entity_type,
            "permission": req.permission,
            "subject": {"type": req.subject.type, "id": req.subject.id},
        }
        if req.subject.relation:
            payload["subject"]["relation"] = req.subject.relation
        if req.metadata:
            payload["metadata"] = req.metadata

        response = await self.client.post(url, json=payload)

        if response.status_code != 200:
            logger.error(f"Lookup entities failed: {response.text}")
            raise Exception(f"Lookup entities failed: {response.status_code}")

        result = response.json()
        return LookupEntityResponse(entity_ids=result.get("entity_ids", []))

    # Payment Switch specific helper methods

    async def can_access_merchant(
        self,
        user_id: str,
        merchant_id: str,
        permission: str = "view"
    ) -> bool:
        """Check if a user can access a merchant"""
        result = await self.check(CheckRequest(
            entity=Entity(type="merchant", id=merchant_id),
            permission=permission,
            subject=Subject(type="user", id=user_id),
        ))
        return result.can

    async def can_access_transaction(
        self,
        user_id: str,
        transaction_id: str,
        permission: str = "view"
    ) -> bool:
        """Check if a user can access a transaction"""
        result = await self.check(CheckRequest(
            entity=Entity(type="transaction", id=transaction_id),
            permission=permission,
            subject=Subject(type="user", id=user_id),
        ))
        return result.can

    async def assign_user_to_merchant(
        self,
        user_id: str,
        merchant_id: str,
        role: str = "member"
    ) -> None:
        """Assign a user to a merchant with a role"""
        await self.write_relationships([
            Tuple(
                entity=Entity(type="merchant", id=merchant_id),
                relation=role,
                subject=Subject(type="user", id=user_id),
            )
        ])

    async def remove_user_from_merchant(
        self,
        user_id: str,
        merchant_id: str,
        role: str = "member"
    ) -> None:
        """Remove a user from a merchant"""
        await self.delete_relationships([
            Tuple(
                entity=Entity(type="merchant", id=merchant_id),
                relation=role,
                subject=Subject(type="user", id=user_id),
            )
        ])

    async def get_user_merchants(
        self,
        user_id: str,
        permission: str = "view"
    ) -> List[str]:
        """Get all merchants a user has access to"""
        result = await self.lookup_entities(LookupEntityRequest(
            entity_type="merchant",
            permission=permission,
            subject=Subject(type="user", id=user_id),
        ))
        return result.entity_ids


# Singleton instance for convenience
_permify_adapter: Optional[PermifyAdapter] = None


def get_permify_adapter() -> PermifyAdapter:
    """Get the global Permify adapter instance"""
    global _permify_adapter
    if _permify_adapter is None:
        import os
        config = PermifyConfig(
            url=os.getenv("PERMIFY_URL", "http://permify:3476"),
            tenant_id=os.getenv("PERMIFY_TENANT_ID", "payment-switch"),
            api_key=os.getenv("PERMIFY_API_KEY"),
        )
        _permify_adapter = PermifyAdapter(config)
    return _permify_adapter


def init_permify_adapter(config: PermifyConfig) -> PermifyAdapter:
    """Initialize the global Permify adapter with custom config"""
    global _permify_adapter
    _permify_adapter = PermifyAdapter(config)
    return _permify_adapter
